src/sqlfluff/rules/L003.py

```diff
@@ -1,15 +1,180 @@
+        }
+
+        return print_dict.__repr__()
+
+    @property
+    def template_content(self):  # pragma: no cover
+        return "".join(
+            seg.raw or getattr(seg, "source_str", "") for seg in self.line_buffer
+        )
+
+    def memory_line_reset(self):
+        """When acting like a MemoryLine this resets state."""
+        self.templated_line = False
+        self.indent_buffer = []
+        # N.B the clean indent depends on the buffer of the line just pased
+        self.clean_indent = _is_clean_indent(self.line_buffer)
+        self.line_buffer = []
+        self.line_indent_stack = []
+        self.indent_size = 0
+        self.hanger_pos = None
+        self.line_anchor = None
+        self.templated_line_type = None
+        return self
+
+    def from_memo_line(self, line_no: int, templated_file: Optional[TemplatedFile]):
+        """Create a final summary from a memo/marker line."""
+        copied_line_buffer = self.line_buffer[:]
+        # Generate our line summary based on the current state of the MemoLine
+        is_comment_line = all(
+            seg.is_type(
+                "whitespace", "comment", "indent"  # dedent is a subtype of indent
+            )
+            for seg in copied_line_buffer
+        )
+        is_empty_line = not any(
+            elem.is_code or elem.is_type("placeholder") for elem in copied_line_buffer
+        )
+        line_summary = self.__class__(
+            line_no=line_no,
+            templated_line=self.templated_line,
+            # Using slicing to copy line_buffer here to be py2 compliant
+            line_buffer=copied_line_buffer,
+            indent_buffer=self.indent_buffer,
+            indent_size=self.indent_size,
+            indent_balance=self.indent_balance,
+            anchor_indent_balance=self.anchor_indent_balance,
+            hanging_indent=self.hanger_pos if self.line_indent_stack else None,
+            # Clean indent is true if the line *ends* with an indent
+            # or has an indent in the initial whitespace.
+            clean_indent=self.clean_indent,
+            # Solidify expensive immutable characteristics
+            templated_line_type=_get_template_block_type(
+                copied_line_buffer, templated_file
+            ),
+            is_comment_line=is_comment_line,
+            is_empty_line=is_empty_line,
+        )
+        # Rest this object so it can continue to be used to track current state
+        self.memory_line_reset()
+        return line_summary
+
+    def set_state_as_of_anchor(
+        self,
+        anchor: Optional[BaseSegment],
+        tab_space_size: int,
+    ):
+        """Create a Line state of this line upon reaching the anchor."""
+        self.anchor_indent_balance = self.indent_balance
+        self.indent_size = _indent_size(
+            self.indent_buffer,
+            tab_space_size=tab_space_size,
+        )
+        self.line_anchor = anchor
+
+        return self
```

Comment:
Do we need this `return` statement? There's no type annotation about it returning anything, and the only place I see it being called, the return value is not used.

Reply:
Were you still going to add the type annotation? TBH, I would prefer to omit the `return` and type hint in this PR, then maybe revisit these things together in a future PR. It's confusing and not useful for the function to return itself -- someone could easily think it's returning a copy.

Reply:
Funny I'd be totally on the other side. Returning and reassigning that return makes it clear what has changed. I'd actually be trying to indicate that this might as well be a new instance - if performance would allow I'd recreate here.

I could move this from being a method to a pure function.